File: packages/python/onnx9000-core/src/onnx9000/core/models/mobilevit.py
"""Mobilevit."""

from typing import Any, Optional

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, flatten, global_average_pool, reshape
from onnx9000.core.primitives import (
    BatchNormalization,
    ConvND,
    DepthwiseConv,
    Gemm,
    LayerNormalization,
    MultiHeadAttention,
    Silu,
)


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:
    """Docstring for D103."""
    return Variable(name=name, shape=shape, dtype=dtype)


class MobileViTBlock:
    """MobileViT block combining Convolutions and Transformers."""

    def __init__(self, dim: int, out_dim: int, num_heads: int, mlp_dim: int, prefix: str = ""):
        """Docstring for D107."""
        self.prefix = prefix
        self.dim = dim
        self.out_dim = out_dim

        # Local representation
        self.local_conv1 = ConvND(2, dim, dim, kernel_size=3, padding=1, bias=False)
        self.local_bn1 = BatchNormalization(dim)
        self.local_act1 = Silu()

        self.local_conv2 = ConvND(2, dim, dim, kernel_size=1, bias=False)
        self.local_bn2 = BatchNormalization(dim)
        self.local_act2 = Silu()

        # Global representation (Transformer)
        self.attn = MultiHeadAttention(num_heads=num_heads)
        self.norm1 = LayerNormalization((dim,))
        self.norm2 = LayerNormalization((dim,))
        self.mlp_fc1 = Gemm(trans_b=1)
        self.mlp_act = Silu()
        self.mlp_fc2 = Gemm(trans_b=1)

        # Fusion
        self.fusion_conv1 = ConvND(2, dim, dim, kernel_size=1, bias=False)
        self.fusion_bn1 = BatchNormalization(dim)
        self.fusion_act1 = Silu()

        self.fusion_conv2 = ConvND(2, dim * 2, out_dim, kernel_size=3, padding=1, bias=False)
        self.fusion_bn2 = BatchNormalization(out_dim)
        self.fusion_act2 = Silu()

    def __call__(self, x: Tensor) -> Tensor:
        """Docstring for D102."""
        identity = x

        # Local
        out = self.local_conv1(
            x,
            get_param(
                f"{self.prefix}.local_conv1.weight",
                [self.local_conv1.out_channels, self.local_conv1.in_channels, 3, 3],
            ),
        )
        out = self.local_bn1(
            out,
            get_param(f"{self.prefix}.local_bn1.weight", [self.dim]),
            get_param(f"{self.prefix}.local_bn1.bias", [self.dim]),
            get_param(f"{self.prefix}.local_bn1.running_mean", [self.dim]),
            get_param(f"{self.prefix}.local_bn1.running_var", [self.dim]),
        )
        out = self.local_act1(out)

        out = self.local_conv2(
            out,
            get_param(
                f"{self.prefix}.local_conv2.weight",
                [self.local_conv2.out_channels, self.local_conv2.in_channels, 1, 1],
            ),
        )
        out = self.local_bn2(
            out,
            get_param(f"{self.prefix}.local_bn2.weight", [self.dim]),
            get_param(f"{self.prefix}.local_bn2.bias", [self.dim]),
            get_param(f"{self.prefix}.local_bn2.running_mean", [self.dim]),
            get_param(f"{self.prefix}.local_bn2.running_var", [self.dim]),
        )
        out = self.local_act2(out)

        # Transformer (assuming unfolded/reshaped for sequence here, simplified for AST building)
        # MobileViT unfolds into patches, but we just use the primitives.
        attn_out = self.norm1(
            out,
            get_param(f"{self.prefix}.norm1.weight", [self.dim]),
            get_param(f"{self.prefix}.norm1.bias", [self.dim]),
        )
        attn_out = self.attn(attn_out, attn_out, attn_out)  # Q, K, V
        attn_out = add(out, attn_out)

        mlp_out = self.norm2(
            attn_out,
            get_param(f"{self.prefix}.norm2.weight", [self.dim]),
            get_param(f"{self.prefix}.norm2.bias", [self.dim]),
        )
        mlp_out = self.mlp_fc1(
            mlp_out,
            get_param(f"{self.prefix}.mlp_fc1.weight", [self.dim * 2, self.dim]),
            get_param(f"{self.prefix}.mlp_fc1.bias", [self.dim * 2]),
        )
        mlp_out = self.mlp_act(mlp_out)
        mlp_out = self.mlp_fc2(
            mlp_out,
            get_param(f"{self.prefix}.mlp_fc2.weight", [self.dim, self.dim * 2]),
            get_param(f"{self.prefix}.mlp_fc2.bias", [self.dim]),
        )
        attn_out = add(attn_out, mlp_out)

        # Fusion
        out = self.fusion_conv1(
            attn_out, get_param(f"{self.prefix}.fusion_conv1.weight", [self.dim, self.dim, 1, 1])
        )
        out = self.fusion_bn1(
            out,
            get_param(f"{self.prefix}.fusion_bn1.weight", [self.dim]),
            get_param(f"{self.prefix}.fusion_bn1.bias", [self.dim]),
            get_param(f"{self.prefix}.fusion_bn1.running_mean", [self.dim]),
            get_param(f"{self.prefix}.fusion_bn1.running_var", [self.dim]),
        )
        out = self.fusion_act1(out)

        from onnx9000.core.ops import concat

        # Concat along channel dim
        out = concat([identity, out], axis=1)

        out = self.fusion_conv2(
            out, get_param(f"{self.prefix}.fusion_conv2.weight", [self.out_dim, self.dim * 2, 3, 3])
        )
        out = self.fusion_bn2(
            out,
            get_param(f"{self.prefix}.fusion_bn2.weight", [self.out_dim]),
            get_param(f"{self.prefix}.fusion_bn2.bias", [self.out_dim]),
            get_param(f"{self.prefix}.fusion_bn2.running_mean", [self.out_dim]),
            get_param(f"{self.prefix}.fusion_bn2.running_var", [self.out_dim]),
        )
        out = self.fusion_act2(out)

        return out


class MobileViT:
    """MobileViT implementation."""

    def __init__(self, num_classes: int = 1000):
        """Docstring for D107."""
        self.num_classes = num_classes
        self.stem_conv = ConvND(2, 3, 16, kernel_size=3, stride=2, padding=1, bias=False)
        self.stem_bn = BatchNormalization(16)
        self.stem_act = Silu()

        self.block1 = MobileViTBlock(16, 32, num_heads=4, mlp_dim=64, prefix="block1")

        self.head_conv = ConvND(2, 32, 1280, kernel_size=1, bias=False)
        self.head_bn = BatchNormalization(1280)
        self.head_act = Silu()

        self.classifier = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:
        """Docstring for D102."""
        x = self.stem_conv(x, get_param("stem_conv.weight", [16, 3, 3, 3]))
        x = self.stem_bn(
            x,
            get_param("stem_bn.weight", [16]),
            get_param("stem_bn.bias", [16]),
            get_param("stem_bn.running_mean", [16]),
            get_param("stem_bn.running_var", [16]),
        )
        x = self.stem_act(x)

        x = self.block1(x)

        x = self.head_conv(x, get_param("head_conv.weight", [1280, 32, 1, 1]))
        x = self.head_bn(
            x,
            get_param("head_bn.weight", [1280]),
            get_param("head_bn.bias", [1280]),
            get_param("head_bn.running_mean", [1280]),
            get_param("head_bn.running_var", [1280]),
        )
        x = self.head_act(x)

        x = global_average_pool(x)
        x = flatten(x)
        x = self.classifier(
            x,
            get_param("classifier.weight", [self.num_classes, 1280]),
            get_param("classifier.bias", [self.num_classes]),
        )

        return x


def mobilevit_s(**kwargs: Any) -> MobileViT:
    """Docstring for D103."""
    return MobileViT(**kwargs)

File: packages/js/core/src/models/mobilevit.ts
import { Tensor } from '../ir/tensor.js';
import {
  BatchNormalization,
  ConvND,
  Gemm,
  LayerNormalization,
  MultiHeadAttention,
  Silu,
} from '../primitives.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class MobileViTBlock {
  public prefix: string;
  public dim: number;
  public outDim: number;

  public localConv1: ConvND;
  public localBn1: BatchNormalization;
  public localAct1: Silu;

  public localConv2: ConvND;
  public localBn2: BatchNormalization;
  public localAct2: Silu;

  public attn: MultiHeadAttention;
  public norm1: LayerNormalization;
  public norm2: LayerNormalization;
  public mlpFc1: Gemm;
  public mlpAct: Silu;
  public mlpFc2: Gemm;

  public fusionConv1: ConvND;
  public fusionBn1: BatchNormalization;
  public fusionAct1: Silu;

  public fusionConv2: ConvND;
  public fusionBn2: BatchNormalization;
  public fusionAct2: Silu;

  constructor(dim: number, outDim: number, numHeads: number, mlpDim: number, prefix: string = '') {
    this.prefix = prefix;
    this.dim = dim;
    this.outDim = outDim;

    this.localConv1 = new ConvND(2, dim, dim, 3, 1, 1, 1, 1, false);
    this.localBn1 = new BatchNormalization(dim);
    this.localAct1 = new Silu();

    this.localConv2 = new ConvND(2, dim, dim, 1, 1, 0, 1, 1, false);
    this.localBn2 = new BatchNormalization(dim);
    this.localAct2 = new Silu();

    this.attn = new MultiHeadAttention(numHeads);
    this.norm1 = new LayerNormalization([dim]);
    this.norm2 = new LayerNormalization([dim]);
    this.mlpFc1 = new Gemm(1.0, 1.0, 0, 1);
    this.mlpAct = new Silu();
    this.mlpFc2 = new Gemm(1.0, 1.0, 0, 1);

    this.fusionConv1 = new ConvND(2, dim, dim, 1, 1, 0, 1, 1, false);
    this.fusionBn1 = new BatchNormalization(dim);
    this.fusionAct1 = new Silu();

    this.fusionConv2 = new ConvND(2, dim * 2, outDim, 3, 1, 1, 1, 1, false);
    this.fusionBn2 = new BatchNormalization(outDim);
    this.fusionAct2 = new Silu();
  }

  call(x: Tensor): Tensor {
    const identity = x;

    let out = this.localConv1.call(
      x,
      getParam(`${this.prefix}.local_conv1.weight`, [
        this.localConv1.outChannels,
        this.localConv1.inChannels,
        3,
        3,
      ]),
    );
    out = this.localBn1.call(
      out,
      getParam(`${this.prefix}.local_bn1.weight`, [this.dim]),
      getParam(`${this.prefix}.local_bn1.bias`, [this.dim]),
      getParam(`${this.prefix}.local_bn1.running_mean`, [this.dim]),
      getParam(`${this.prefix}.local_bn1.running_var`, [this.dim]),
    );
    out = this.localAct1.call(out);

    out = this.localConv2.call(
      out,
      getParam(`${this.prefix}.local_conv2.weight`, [
        this.localConv2.outChannels,
        this.localConv2.inChannels,
        1,
        1,
      ]),
    );
    out = this.localBn2.call(
      out,
      getParam(`${this.prefix}.local_bn2.weight`, [this.dim]),
      getParam(`${this.prefix}.local_bn2.bias`, [this.dim]),
      getParam(`${this.prefix}.local_bn2.running_mean`, [this.dim]),
      getParam(`${this.prefix}.local_bn2.running_var`, [this.dim]),
    );
    out = this.localAct2.call(out);

    let attnOut = this.norm1.call(
      out,
      getParam(`${this.prefix}.norm1.weight`, [this.dim]),
      getParam(`${this.prefix}.norm1.bias`, [this.dim]),
    );
    attnOut = this.attn.call(attnOut, attnOut, attnOut);
    attnOut = recordOp('Add', [out, attnOut]);

    let mlpOut = this.norm2.call(
      attnOut,
      getParam(`${this.prefix}.norm2.weight`, [this.dim]),
      getParam(`${this.prefix}.norm2.bias`, [this.dim]),
    );
    mlpOut = this.mlpFc1.call(
      mlpOut,
      getParam(`${this.prefix}.mlp_fc1.weight`, [this.dim * 2, this.dim]),
      getParam(`${this.prefix}.mlp_fc1.bias`, [this.dim * 2]),
    );
    mlpOut = this.mlpAct.call(mlpOut);
    mlpOut = this.mlpFc2.call(
      mlpOut,
      getParam(`${this.prefix}.mlp_fc2.weight`, [this.dim, this.dim * 2]),
      getParam(`${this.prefix}.mlp_fc2.bias`, [this.dim]),
    );
    attnOut = recordOp('Add', [attnOut, mlpOut]);

    out = this.fusionConv1.call(
      attnOut,
      getParam(`${this.prefix}.fusion_conv1.weight`, [this.dim, this.dim, 1, 1]),
    );
    out = this.fusionBn1.call(
      out,
      getParam(`${this.prefix}.fusion_bn1.weight`, [this.dim]),
      getParam(`${this.prefix}.fusion_bn1.bias`, [this.dim]),
      getParam(`${this.prefix}.fusion_bn1.running_mean`, [this.dim]),
      getParam(`${this.prefix}.fusion_bn1.running_var`, [this.dim]),
    );
    out = this.fusionAct1.call(out);

    out = recordOp('Concat', [identity, out], { axis: 1 });

    out = this.fusionConv2.call(
      out,
      getParam(`${this.prefix}.fusion_conv2.weight`, [this.outDim, this.dim * 2, 3, 3]),
    );
    out = this.fusionBn2.call(
      out,
      getParam(`${this.prefix}.fusion_bn2.weight`, [this.outDim]),
      getParam(`${this.prefix}.fusion_bn2.bias`, [this.outDim]),
      getParam(`${this.prefix}.fusion_bn2.running_mean`, [this.outDim]),
      getParam(`${this.prefix}.fusion_bn2.running_var`, [this.outDim]),
    );
    out = this.fusionAct2.call(out);

    return out;
  }
}

export class MobileViT {
  public numClasses: number;
  public stemConv: ConvND;
  public stemBn: BatchNormalization;
  public stemAct: Silu;
  public block1: MobileViTBlock;
  public headConv: ConvND;
  public headBn: BatchNormalization;
  public headAct: Silu;
  public classifier: Gemm;

  constructor(numClasses: number = 1000) {
    this.numClasses = numClasses;
    this.stemConv = new ConvND(2, 3, 16, 3, 2, 1, 1, 1, false);
    this.stemBn = new BatchNormalization(16);
    this.stemAct = new Silu();

    this.block1 = new MobileViTBlock(16, 32, 4, 64, 'block1');

    this.headConv = new ConvND(2, 32, 1280, 1, 1, 0, 1, 1, false);
    this.headBn = new BatchNormalization(1280);
    this.headAct = new Silu();

    this.classifier = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    x = this.stemConv.call(x, getParam('stem_conv.weight', [16, 3, 3, 3]));
    x = this.stemBn.call(
      x,
      getParam('stem_bn.weight', [16]),
      getParam('stem_bn.bias', [16]),
      getParam('stem_bn.running_mean', [16]),
      getParam('stem_bn.running_var', [16]),
    );
    x = this.stemAct.call(x);

    x = this.block1.call(x);

    x = this.headConv.call(x, getParam('head_conv.weight', [1280, 32, 1, 1]));
    x = this.headBn.call(
      x,
      getParam('head_bn.weight', [1280]),
      getParam('head_bn.bias', [1280]),
      getParam('head_bn.running_mean', [1280]),
      getParam('head_bn.running_var', [1280]),
    );
    x = this.headAct.call(x);

    x = recordOp('GlobalAveragePool', [x]);
    x = recordOp('Flatten', [x]);
    x = this.classifier.call(
      x,
      getParam('classifier.weight', [this.numClasses, 1280]),
      getParam('classifier.bias', [this.numClasses]),
    );

    return x;
  }
}

export function mobilevitS(numClasses: number = 1000): MobileViT {
  return new MobileViT(numClasses);
}
